relay-server/src/services/server/io.rs

```diff
@@ -0,0 +1,265 @@
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use futures::FutureExt;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::time::{Instant, Sleep};
+
+use crate::statsd::RelayCounters;
+
+/// A wrapper for [`AsyncRead`] and [`AsyncWrite`] streams with a maximum idle time.
+///
+/// If there is no activity in the underlying stream for the specified `timeout`
+/// the [`IdleTimeout`] will abort the operation and return [`std::io::ErrorKind::TimedOut`].
+pub struct IdleTimeout<T> {
+    inner: T,
+    timeout: Option<Duration>,
```

Comment:
Not sure if this would make typing annoying, but it would be cool if `timeout` was non-optional here. So if there's no timeout set in the config, don't use `IdleTimeout`.